### PR TITLE
SALTO-1830 Post deploy message

### PR DESCRIPTION
**Pull request by rotem531**
added post deployment summary to the cli
image
image
image
image

---

when deployment isn't fully successful, post deployment summary will be printed listing which elements where deployed successfully, which were partially successful and which have failed

---
_Release Notes_: 
CLI:
added post deployment summary

---
_User Notifications_: 
none

### CODE REVIEW

File: packages/cli/src/commands/deploy.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ import { logger } from '@salto-io/logging'
 import { Workspace } from '@salto-io/workspace'
 import { mkdirp, writeFile } from '@salto-io/file'
 import path from 'path'
+import { DetailedChangeId, DeploySummaryResult } from '@salto-io/core/src/core/deploy'
```
Comment (Uri-Bar) [inline packages/cli/src/commands/deploy.ts:24]:
```suggestion
import { DetailedChangeId, DeploySummaryResult } from '@salto-io/core'
```
1. import from packages should be at the top (above path)
2. when we import from package we dont write the exact path - see the others imports above. so we need to make sure that core is actually export this types

File: packages/cli/src/commands/deploy.ts [hunk 2]
```diff
@@ -78,6 +80,24 @@ const printStartDeploy = async (output: CliOutput, executingDeploy: boolean, che
...
+): Promise<void> => {
+  const summaryResultToId = Object.entries(summary).reduce(
+    (acc, [key, value]) => {
+      acc[value] = acc[value] || []
+      acc[value].push(key)
+      return acc
+    },
+    {} as Record<DeploySummaryResult, DetailedChangeId[]>,
+  )
+  const formattedDeploymentSummary: string | undefined = formatDeploymentSummary(summaryResultToId)
```
Comment (Uri-Bar) [inline packages/cli/src/commands/deploy.ts:95]:
```suggestion
  const formattedDeploymentSummary = formatDeploymentSummary(summaryResultToId)
```

File: packages/cli/src/commands/deploy.ts [hunk 3]
```diff
@@ -78,6 +80,24 @@ const printStartDeploy = async (output: CliOutput, executingDeploy: boolean, che
...
+  summary: Record<DetailedChangeId, DeploySummaryResult>,
+  output: CliOutput,
+): Promise<void> => {
+  const summaryResultToId = Object.entries(summary).reduce(
+    (acc, [key, value]) => {
+      acc[value] = acc[value] || []
+      acc[value].push(key)
+      return acc
+    },
+    {} as Record<DeploySummaryResult, DetailedChangeId[]>,
```
Comment (Uri-Bar) [inline packages/cli/src/commands/deploy.ts:93]:
```suggestion
  const summaryResultToId: Record<DeploySummaryResult, DetailedChangeId[]> = Object.entries(summary).reduce(
    (acc, [key, value]) => {
      acc[value] = acc[value] || []
      acc[value].push(key)
      return acc
    },
```
you should avoid using casting here, maybe try adding the type in the declaration or inside the iterator

File: packages/cli/src/commands/deploy.ts [hunk 4]
```diff
@@ -268,11 +288,13 @@ export const action: WorkspaceCommandAction<DeployArgs> = async ({
     changeError =>
       summary[changeError.elemID.getFullName()] !== 'failure' || changeError.deployActions?.postAction?.showOnFailure,
   )
+  await printDeploymentSummary(summary, output)
 
   const postDeployActionsOutput = formatDeployActions({
     wsChangeErrors: changeErrorsForPostDeployOutput,
     isPreDeploy: false,
   })
+
```
Comment (Uri-Bar) [inline packages/cli/src/commands/deploy.ts:297]: no need

File: packages/cli/src/formatter.ts [hunk 5]
```diff
@@ -9,6 +9,7 @@ import _ from 'lodash'
 import { EOL } from 'os'
 import chalk from 'chalk'
 import wu, { WuIterable } from 'wu'
+import { DetailedChangeId, DeploySummaryResult } from '@salto-io/core/src/core/deploy'
```
Comment (Uri-Bar) [inline packages/cli/src/formatter.ts:12]: same comment on the import 1. placement 2. path

File: packages/cli/src/commands/deploy.ts (hunk 3 above)
Comment (Uri-Bar) [reply packages/cli/src/commands/deploy.ts:93]: I actually think that we can get this info from summarizeDeployChanges.
something like:
export const summarizeDeployChanges = (
  requested: Change[],
  applied: Change[],
): { a: Record<DetailedChangeId, DeploySummaryResult>; b: Record<DeploySummaryResult, DetailedChangeId[]> } => {
  const appliedById = _.keyBy(applied, c => getChangeData(c).elemID.getFullName())
  const a = Object.fromEntries(
    requested.flatMap(requestedChange => {
      const requestedChangeId = getChangeData(requestedChange).elemID.getFullName()
      return summarizeDeployChange(requestedChange, appliedById?.[requestedChangeId])
    }),
  )
  const b: Record<DeploySummaryResult, DetailedChangeId[]> = {
    success: [],
    failure: [],
    'partial-success': [],
  }

  Object.entries(a).forEach(([changeId, resultValue]) => {
    if (b[resultValue]) {
      b[resultValue].push(changeId)
    }
  })
  return { a, b }
}

File: packages/cli/src/formatter.ts [hunk 6]
```diff
@@ -326,6 +327,47 @@ export const formatDeployActions = ({
...
 }
 
+export const formatDeploymentSummaryResult = (ids: DetailedChangeId[], result: DeploySummaryResult): string => {
+  const mapping: Record<DeploySummaryResult, DetailedChangeId | null> = {
+    failure: Prompts.DEPLOYMENT_STATUS.failure,
+    'partial-success': Prompts.DEPLOYMENT_STATUS.partialSuccess,
+    success: Prompts.DEPLOYMENT_STATUS.success,
+  }
+
+  const sign: string | null = mapping[result]
```
Comment (Uri-Bar) [inline packages/cli/src/formatter.ts:337]:
```suggestion
  const sign = mapping[result]
```
you dont need to state the types everywhere, typescript know most of the types by itself. so Im not commenting on all of them but remove them

File: packages/cli/src/prompts.ts [hunk 7]
```diff
@@ -55,6 +55,13 @@ export default class Prompts {
     eq: '|',
   }
 
+  public static readonly DEPLOYMENT_STATUS = {
+    partialSuccess: chalk.yellow.bold('P'),
```
Comment (Uri-Bar) [inline packages/cli/src/prompts.ts:59]: no need for bold in all 3

File: packages/cli/src/prompts.ts [hunk 8]
```diff
@@ -90,6 +97,13 @@ export default class Prompts {
   public static readonly DID_YOU_MEAN = 'Did you mean'
   public static readonly DESCRIBE_NOT_FOUND = 'Unknown element type.'
 
+  public static readonly DEPLOYMENT_SUMMARY_HEADLINE = chalk.bgBlack.bold('Deployment summary:')
+  public static readonly EXPLAIN_DEPLOYMENT_SUMMARY =
```
Comment (Uri-Bar) [inline packages/cli/src/prompts.ts:101]:
```suggestion
  public static readonly DEPLOYMENT_SUMMARY_LAGEND =
```
check the spelling

File: packages/cli/src/formatter.ts [hunk 9]
```diff
@@ -326,6 +327,47 @@ export const formatDeployActions = ({
...
+  const succeeded: string | null = Object.prototype.hasOwnProperty.call(summary, successResult)
+    ? formatDeploymentSummaryResult(summary[successResult], successResult)
+    : null
+  const partiallySucceeded: string | null = Object.prototype.hasOwnProperty.call(summary, partialSuccessResult)
+    ? formatDeploymentSummaryResult(summary[partialSuccessResult], partialSuccessResult)
+    : null
+  const failed: string | null = Object.prototype.hasOwnProperty.call(summary, failureResult)
+    ? formatDeploymentSummaryResult(summary[failureResult], failureResult)
+    : null
+  return !Object.prototype.hasOwnProperty.call(summary, failureResult) &&
```
Comment (Uri-Bar) [inline packages/cli/src/formatter.ts:358]: why arent you using the partiallySucceeded and failed vars?

File: packages/cli/test/commands/deploy.test.ts [hunk 10]
```diff
@@ -492,4 +498,84 @@ describe('deploy command', () => {
       expect(workspace.setCurrentEnv).toHaveBeenCalledWith(mocks.withEnvironmentParam, false)
     })
   })
+  describe('Post deployment summary', () => {
```
Comment (Uri-Bar) [inline packages/cli/test/commands/deploy.test.ts:501]: nice tests!

**Review by Uri-Bar — COMMENTED** [review]
looks great, added few comments.
plus:
1. in the title use all caps - SALTO
2. in the release notes change it to 
CLI:
print post deployment summary

**Comment by Uri-Bar** [comment]
please check what happens when it all failed and stating that all of the elements failed.
plus, add uts

File: packages/cli/test/commands/deploy.test.ts [hunk 11]
```diff
@@ -492,4 +498,84 @@ describe('deploy command', () => {
...
+    describe('when deployment is successful', () => {
```
Comment (Uri-Bar) [inline packages/cli/test/commands/deploy.test.ts:502]:
```suggestion
    describe('when all of the deployment changes are successful', () => {
```

File: packages/cli/test/commands/deploy.test.ts [hunk 12]
```diff
@@ -492,4 +498,84 @@ describe('deploy command', () => {
...
+            force: false,
+            dryRun: false,
+            detailedPlan: true,
+            checkOnly: false,
+            accounts,
+          },
+          workspace,
+        })
+      })
+      it('should not print anything', async () => {
```
Comment (Uri-Bar) [inline packages/cli/test/commands/deploy.test.ts:520]:
```suggestion
      it('should not print post summary', async () => {
```

File: packages/cli/test/commands/deploy.test.ts [hunk 13]
```diff
@@ -492,4 +498,84 @@ describe('deploy command', () => {
...
+        expect(output.stdout.content).not.toContain(Prompts.DEPLOYMENT_SUMMARY_HEADLINE)
+        expect(output.stdout.content).not.toContain(chalk.green.bold('S'))
+        expect(output.stdout.content).not.toContain(chalk.yellow.bold('P'))
+        expect(output.stdout.content).not.toContain(chalk.red.bold('F'))
+      })
+    })
+    describe('when deployment failed', () => {
```
Comment (Uri-Bar) [inline packages/cli/test/commands/deploy.test.ts:527]: like the success comment

File: packages/cli/test/commands/deploy.test.ts [hunk 14]
```diff
@@ -492,4 +498,84 @@ describe('deploy command', () => {
...
+          workspace,
+        })
+      })
+      it('should print all failed elements as failed', async () => {
+        expect(output.stdout.content).toContain(Prompts.DEPLOYMENT_SUMMARY_HEADLINE)
+        expect(output.stdout.content).toContain(`${chalk.red.bold('F')} instance_test`)
+        expect(output.stdout.content).toContain(`${chalk.red.bold('F')} test_instance`)
+      })
+      it('should not print succeeded or partially succeeded', async () => {
+        expect(output.stdout.content).not.toContain(chalk.green.bold('S'))
```
Comment (Uri-Bar) [inline packages/cli/test/commands/deploy.test.ts:551]: arent this print in the legend anyway?

File: packages/cli/test/commands/deploy.test.ts [hunk 15]
```diff
@@ -492,4 +498,84 @@ describe('deploy command', () => {
...
+        expect(output.stdout.content).not.toContain(chalk.yellow.bold('P'))
+      })
+    })
+    describe('when deployment is partially successful', () => {
```
Comment (Uri-Bar) [inline packages/cli/test/commands/deploy.test.ts:555]:
```suggestion
    describe('when some of the deployment changes are partially successful', () => {
```

File: packages/cli/src/formatter.ts (hunk 9 above)
Comment (rotem531) [reply packages/cli/src/formatter.ts:358]: im not sure i understand
im using them at the final part of the return:
[succeeded, partiallySucceeded, failed].filter(item => item != null).join('\n'),

File: packages/cli/test/commands/deploy.test.ts (hunk 15 above)
Comment (rotem531) [reply packages/cli/test/commands/deploy.test.ts:555]: this use case is not only for when sone elements are partially successful, but when some are successful and some are not, or everything is partially successful
i fixed it with more specific UT

File: packages/cli/src/commands/deploy.ts (hunk 4 above)
Comment (Uri-Bar) [reply packages/cli/src/commands/deploy.ts:283]: still no need

File: packages/cli/src/commands/deploy.ts [hunk 16]
```diff
@@ -263,16 +264,23 @@ export const action: WorkspaceCommandAction<DeployArgs> = async ({
...
-  const summary = summarizeDeployChanges(requested, result.appliedChanges ?? [])
+  const { elemIdToResult, resultToElemId } = summarizeDeployChanges(requested, result.appliedChanges ?? [])
   const changeErrorsForPostDeployOutput = actionPlan.changeErrors.filter(
     changeError =>
-      summary[changeError.elemID.getFullName()] !== 'failure' || changeError.deployActions?.postAction?.showOnFailure,
+      elemIdToResult[changeError.elemID.getFullName()] !== 'failure' ||
+      changeError.deployActions?.postAction?.showOnFailure,
   )
 
+  const formattedDeploymentSummary = formatDeploymentSummary(resultToElemId)
```
Comment (Uri-Bar) [inline packages/cli/src/commands/deploy.ts:274]: looks much cleaner

File: packages/cli/src/commands/deploy.ts [hunk 17]
```diff
@@ -263,16 +264,23 @@ export const action: WorkspaceCommandAction<DeployArgs> = async ({
...
+  if (formattedDeploymentSummary !== undefined) {
```
Comment (Uri-Bar) [inline packages/cli/src/commands/deploy.ts:275]:
```suggestion
  if (!formattedDeploymentSummary) {
```

File: packages/cli/src/prompts.ts [hunk 18]
```diff
@@ -90,6 +97,15 @@ export default class Prompts {
   public static readonly DID_YOU_MEAN = 'Did you mean'
   public static readonly DESCRIBE_NOT_FOUND = 'Unknown element type.'
 
+  public static readonly DEPLOYMENT_SUMMARY_HEADLINE = chalk.bgBlack.bold('Deployment summary:')
+  public static readonly ALL_DEPLOYMENT_ELEMENTS_FAILED = `${chalk.red('X')} All elements failed deployment`
+  public static readonly ALL_DEPLOYMENT_ELEMENTS_SUCCEEDED = `${chalk.green('\u2713')} All elements succeeded deployment`
```
Comment (Uri-Bar) [inline packages/cli/src/prompts.ts:102]:
```suggestion
  public static readonly ALL_DEPLOYMENT_ELEMENTS_SUCCEEDED = `${chalk.green('\u2713')} All elements were successfully deployed`
```

File: packages/cli/src/prompts.ts [hunk 19]
```diff
@@ -90,6 +97,15 @@ export default class Prompts {
   public static readonly DID_YOU_MEAN = 'Did you mean'
   public static readonly DESCRIBE_NOT_FOUND = 'Unknown element type.'
 
+  public static readonly DEPLOYMENT_SUMMARY_HEADLINE = chalk.bgBlack.bold('Deployment summary:')
+  public static readonly ALL_DEPLOYMENT_ELEMENTS_FAILED = `${chalk.red('X')} All elements failed deployment`
```
Comment (Uri-Bar) [inline packages/cli/src/prompts.ts:101]:
```suggestion
  public static readonly ALL_DEPLOYMENT_ELEMENTS_FAILED = `${chalk.red('X')} All elements failed to deploy`
```

**Review by Uri-Bar — COMMENTED** [review]
very small comments

**Review by Uri-Bar — APPROVED** [review]
Good job

**Pull request merged**